Bottom: d2effc0fadc8275ddb0bcb2e1200ef61b1eac387
Top:    190d2bc2df0108ff18a7816d6406e7b2ebcbf24d
Author: Peter Clifton <pcjc2@cam.ac.uk>
Date:   2009-02-06 17:24:15 +0000

Add colour information to VBO





---

diff --git a/src/hid/common/hidgl.c b/src/hid/common/hidgl.c
index ccf74af..5e88503 100644
--- a/src/hid/common/hidgl.c
+++ b/src/hid/common/hidgl.c
@@ -48,6 +48,10 @@
 RCSID ("$Id: $");
 
 triangle_buffer buffer;
+GLfloat cur_r = 1.;
+GLfloat cur_g = 1.;
+GLfloat cur_b = 1.;
+GLfloat cur_a = 1.;
 
 #if 0
 triangle_array *
@@ -64,7 +68,8 @@ hidgl_init_triangle_array (triangle_buffer *buffer)
   const GLubyte *errString;
 
   buffer->triangle_count = 0;
-  buffer->coord_comp_count = 0;
+//  buffer->coord_comp_count = 0;
+  buffer->array_size = TRIANGLE_ARRAY_BYTES / (3 * sizeof (tri_array_element));
 
   glEnableClientState (GL_VERTEX_ARRAY);
   glGenBuffers (1, &buffer->vbo_name);
@@ -92,8 +97,11 @@ hidgl_flush_triangles (triangle_buffer *buffer)
 
   glUnmapBuffer (GL_ARRAY_BUFFER);
 
-  glVertexPointer (2, GL_FLOAT, 0, NULL); // buffer->triangle_array);
+  glEnableClientState (GL_COLOR_ARRAY);
+  glVertexPointer (2, GL_FLOAT, sizeof (tri_array_element), &((tri_array_element *)NULL)->x);
+  glColorPointer (4, GL_FLOAT, sizeof (tri_array_element), &((tri_array_element *)NULL)->r );
   glDrawArrays (GL_TRIANGLES, 0, buffer->triangle_count * 3);
+  glDisableClientState (GL_COLOR_ARRAY);
 
 //  buffer->triangle_count = 0;
 //  buffer->coord_comp_count = 0;
@@ -107,17 +115,30 @@ hidgl_flush_triangles (triangle_buffer *buffer)
 void
 hidgl_ensure_triangle_space (triangle_buffer *buffer, int count)
 {
-  if (count > TRIANGLE_ARRAY_SIZE)
+  if (count > buffer->array_size)
     {
       fprintf (stderr, "Not enough space in vertex buffer\n");
       fprintf (stderr, "Requested %i triangles, %i available\n",
-                       count, TRIANGLE_ARRAY_SIZE);
+                       count, buffer->array_size);
       exit (1);
     }
-  if (count > TRIANGLE_ARRAY_SIZE - buffer->triangle_count)
+  if (count > buffer->array_size - buffer->triangle_count)
     hidgl_flush_triangles (buffer);
 }
 
+
+void
+hidgl_color (GLfloat r, GLfloat g, GLfloat b, GLfloat a)
+{
+  cur_r = r;
+  cur_g = g;
+  cur_b = b;
+  cur_a = a;
+
+  /* We still draw non-filled rectangles in immediate mode */
+  glColor4f (r, g, b, a);
+}
+
 //static int cur_mask = -1;
 
 
@@ -314,7 +335,7 @@ static void draw_cap (double width, int x, int y, double angle, double scale)
   for (i = 0; i < slices; i++) {
     capx =  radius * cosf (angle * M_PI / 180. + ((float)(i + 1)) * M_PI / (float)slices) + x;
     capy = -radius * sinf (angle * M_PI / 180. + ((float)(i + 1)) * M_PI / (float)slices) + y;
-    hidgl_add_triangle (&buffer, last_capx, last_capy, capx, capy, x, y);
+    hidgl_add_triangle (&buffer, last_capx, last_capy, capx, capy, x, y, cur_r, cur_g, cur_b, cur_a);
     last_capx = capx;
     last_capy = capy;
   }
@@ -375,10 +396,10 @@ hidgl_draw_line (int cap, double width, int x1, int y1, int x2, int y2, double s
   hidgl_ensure_triangle_space (&buffer, 2);
   hidgl_add_triangle (&buffer, x1 - wdx, y1 - wdy,
                                x2 - wdx, y2 - wdy,
-                               x2 + wdx, y2 + wdy);
+                               x2 + wdx, y2 + wdy, cur_r, cur_g, cur_b, cur_a);
   hidgl_add_triangle (&buffer, x1 - wdx, y1 - wdy,
                                x2 + wdx, y2 + wdy,
-                               x1 + wdx, y1 + wdy);
+                               x1 + wdx, y1 + wdy, cur_r, cur_g, cur_b, cur_a);
   if (circular_caps)
     {
       draw_cap (width, x1, y1, angle, scale);
@@ -442,10 +463,10 @@ hidgl_draw_arc (double width, int x, int y, int rx, int ry,
     outer_x = -outer_r * cos_ang + x;  outer_y = outer_r * sin_ang + y;
     hidgl_add_triangle (&buffer, last_inner_x, last_inner_y,
                                  last_outer_x, last_outer_y,
-                                 outer_x, outer_y);
+                                 outer_x, outer_y, cur_r, cur_g, cur_b, cur_a);
     hidgl_add_triangle (&buffer, last_inner_x, last_inner_y,
                                  inner_x, inner_y,
-                                 outer_x, outer_y);
+                                 outer_x, outer_y, cur_r, cur_g, cur_b, cur_a);
     last_inner_x = inner_x;  last_inner_y = inner_y;
     last_outer_x = outer_x;  last_outer_y = outer_y;
   }
@@ -497,7 +518,7 @@ hidgl_fill_circle (int vx, int vy, int vr, double scale)
     float x, y;
     x = radius * cosf (((float)(i + 1)) * 2. * M_PI / (float)slices) + vx;
     y = radius * sinf (((float)(i + 1)) * 2. * M_PI / (float)slices) + vy;
-    hidgl_add_triangle (&buffer, vx, vy, last_x, last_y, x, y);
+    hidgl_add_triangle (&buffer, vx, vy, last_x, last_y, x, y, cur_r, cur_g, cur_b, cur_a);
     last_x = x;
     last_y = y;
   }
@@ -584,7 +605,7 @@ myVertex (GLdouble *vertex_data)
           hidgl_add_triangle (&buffer,
                               triangle_vertices [0], triangle_vertices [1],
                               triangle_vertices [2], triangle_vertices [3],
-                              vertex_data [0], vertex_data [1]);
+                              vertex_data [0], vertex_data [1], cur_r, cur_g, cur_b, cur_a);
 
           if (tessVertexType == GL_TRIANGLE_STRIP)
             {
@@ -608,7 +629,7 @@ myVertex (GLdouble *vertex_data)
           hidgl_add_triangle (&buffer,
                               triangle_vertices [0], triangle_vertices [1],
                               triangle_vertices [2], triangle_vertices [3],
-                              triangle_vertices [4], triangle_vertices [5]);
+                              triangle_vertices [4], triangle_vertices [5], cur_r, cur_g, cur_b, cur_a);
           triangle_comp_idx = 0;
           stashed_vertices = 0;
         }
@@ -658,8 +679,8 @@ hidgl_fill_polygon (int n_coords, int *x, int *y)
 void
 hidgl_fill_rect (int x1, int y1, int x2, int y2)
 {
-  hidgl_add_triangle (&buffer, x1, y1, x1, y2, x2, y2);
-  hidgl_add_triangle (&buffer, x2, y1, x2, y2, x1, y1);
+  hidgl_add_triangle (&buffer, x1, y1, x1, y2, x2, y2, cur_r, cur_g, cur_b, cur_a);
+  hidgl_add_triangle (&buffer, x2, y1, x2, y2, x1, y1, cur_r, cur_g, cur_b, cur_a);
 }
 
 /* ---------------------------------------------------------------------- */
diff --git a/src/hid/common/hidgl.h b/src/hid/common/hidgl.h
index f82056a..f5dc1fc 100644
--- a/src/hid/common/hidgl.h
+++ b/src/hid/common/hidgl.h
@@ -26,15 +26,28 @@
 #define __HIDGL_INCLUDED__
 
 //#define TRIANGLE_ARRAY_SIZE 5000
-#define TRIANGLE_ARRAY_SIZE 5461
+//#define TRIANGLE_ARRAY_SIZE 5461
 /* Assumes GLFloat is 4 bytes, and we have X,Y coords x3 for each triangle:
    4 * 5461 * 2 * 3 = 109464 */
-#define TRIANGLE_ARRAY_BYTES 131072
+// #define TRIANGLE_ARRAY_BYTES 131072
+
+typedef struct {
+  GLfloat x;
+  GLfloat y;
+  GLfloat r;
+  GLfloat g;
+  GLfloat b;
+  GLfloat a;
+} tri_array_element;
+
+#define TRIANGLE_ARRAY_BYTES 256 * 1024
 typedef struct {
 //  GLfloat triangle_array [2 * 3 * TRIANGLE_ARRAY_SIZE];
-  GLfloat *triangle_array;
+//  GLfloat *triangle_array;
+  tri_array_element *triangle_array;
   unsigned int triangle_count;
-  unsigned int coord_comp_count;
+//  unsigned int coord_comp_count;
+  unsigned int array_size;
   GLuint vbo_name;
 } triangle_buffer;
 
@@ -43,19 +56,37 @@ extern triangle_buffer buffer;
 void hidgl_init_triangle_array (triangle_buffer *buffer);
 void hidgl_flush_triangles (triangle_buffer *buffer);
 void hidgl_ensure_triangle_space (triangle_buffer *buffer, int count);
+void hidgl_color (GLfloat r, GLfloat g, GLfloat b, GLfloat a);
 
 static inline void
 hidgl_add_triangle (triangle_buffer *buffer,
                     GLfloat x1, GLfloat y1,
                     GLfloat x2, GLfloat y2,
-                    GLfloat x3, GLfloat y3)
+                    GLfloat x3, GLfloat y3,
+                    GLfloat r, GLfloat g,
+                    GLfloat b, GLfloat a)
 {
-  buffer->triangle_array [buffer->coord_comp_count++] = x1;
-  buffer->triangle_array [buffer->coord_comp_count++] = y1;
-  buffer->triangle_array [buffer->coord_comp_count++] = x2;
-  buffer->triangle_array [buffer->coord_comp_count++] = y2;
-  buffer->triangle_array [buffer->coord_comp_count++] = x3;
-  buffer->triangle_array [buffer->coord_comp_count++] = y3;
+  int i = 0;
+  buffer->triangle_array[buffer->triangle_count * 3 + i].x = x1;
+  buffer->triangle_array[buffer->triangle_count * 3 + i].y = y1;
+  buffer->triangle_array[buffer->triangle_count * 3 + i].r = r;
+  buffer->triangle_array[buffer->triangle_count * 3 + i].g = g;
+  buffer->triangle_array[buffer->triangle_count * 3 + i].b = b;
+  buffer->triangle_array[buffer->triangle_count * 3 + i].a = a;
+  i++;
+  buffer->triangle_array[buffer->triangle_count * 3 + i].x = x2;
+  buffer->triangle_array[buffer->triangle_count * 3 + i].y = y2;
+  buffer->triangle_array[buffer->triangle_count * 3 + i].r = r;
+  buffer->triangle_array[buffer->triangle_count * 3 + i].g = g;
+  buffer->triangle_array[buffer->triangle_count * 3 + i].b = b;
+  buffer->triangle_array[buffer->triangle_count * 3 + i].a = a;
+  i++;
+  buffer->triangle_array[buffer->triangle_count * 3 + i].x = x3;
+  buffer->triangle_array[buffer->triangle_count * 3 + i].y = y3;
+  buffer->triangle_array[buffer->triangle_count * 3 + i].r = r;
+  buffer->triangle_array[buffer->triangle_count * 3 + i].g = g;
+  buffer->triangle_array[buffer->triangle_count * 3 + i].b = b;
+  buffer->triangle_array[buffer->triangle_count * 3 + i].a = a;
   buffer->triangle_count++;
 }
 
diff --git a/src/hid/gtk/gtkhid-main.c b/src/hid/gtk/gtkhid-main.c
index 24b499d..90604c2 100644
--- a/src/hid/gtk/gtkhid-main.c
+++ b/src/hid/gtk/gtkhid-main.c
@@ -811,8 +811,7 @@ ghid_set_color (hidGC gc, const char *name)
   if( ! ghid_gui_is_up )
     return;
 
-  hidgl_flush_triangles (&buffer);
-  glColor4d (r, g, b, a);
+  hidgl_color (r, g, b, a);
 }
 
 void
